src/lava/magma/core/model/py/ports.py

```diff
@@ -9,96 +9,229 @@
 from lava.magma.compiler.channels.interfaces import AbstractCspPort
 from lava.magma.compiler.channels.pypychannel import CspSendPort, CspRecvPort
 from lava.magma.core.model.interfaces import AbstractPortImplementation
+from lava.magma.core.model.model import AbstractProcessModel
 from lava.magma.runtime.mgmt_token_enums import enum_to_np, enum_equal
 
 
 class AbstractPyPort(AbstractPortImplementation):
+    """Abstract class for Ports implemented in Python.
+
+    Ports at the Process level provide an interface to connect
+    Processes with each other. PyPorts provide an interface to send and
+    receive messages via channels implemented by a backend messaging
+    infrastructure. The Communicating Sequential Processes (CSP) paradigm is the
+    inspiration of the messaging infrastructure. Thus, a channel denotes a CSP
+    channel of the messaging infrastructure and CSP Ports denote the low level
+    ports also used in the messaging infrastructure. PyPorts are the
+    implementation for message exchange in Python, using the low level CSP Ports
+    of the backend messaging infrastructure. PyPorts provide interface for send
+    and receive of input and output Ports. A PyPort may have one or multiple
+    connection to other PyPorts. These connections are represented by csp_ports,
+    which is a list of CSP ports corresponding to the connected PyPorts.
+    """
     @property
```

Comment:
Some suggestions:

Ports at the Process level provide an interface to connect Processes with each other. **Once two Processes have been connected by Ports, they can exchange data. \newline
Lava provides four types of Ports: _briefly describe the four Port types and what they can do_.\newline
The exchanged data can be of either of four types: _briefly explain the difference between VEC_DENSE, VEC_SPARSE , ..._.\newline
To exchange data,** PyPorts provide an interface to send and
    receive messages via channels implemented by a backend messaging
    infrastructure**, which has been inspired by the** Communicating Sequential Processes (CSP) paradigm ~~is the
    inspiration of the messaging infrastructure~~. Thus, a channel denotes a CSP
    channel of the messaging infrastructure and CSP Ports denote the low level
    ports also used in the messaging infrastructure. PyPorts are the
    implementation for message exchange in Python, using the low level CSP Ports
    of the backend messaging infrastructure. ~~PyPorts provide interface for send
    and receive of input and output Ports.~~ A PyPort may have one or multiple
    connection to other PyPorts. These connections are represented by csp_ports,
    which is a list of CSP ports corresponding to the connected PyPorts.

Reply:
Thanks for the suggestions, I used some of it. I will not explain VEC_DENSE etc. here, because these types are independent of the abstract class and vice versa -> they can change without adapting this class at all.
Same thing for shape.

Reply:
Looks great, thanks!